src/peergos/server/tests/simulation/Simulator.java

```diff
@@ -376,35 +410,72 @@ private void run(final Simulation simulation, final String user) {
+                Optional<Path> path = randomSharedPath.apply(owner, sharee);
+                if (path.isPresent()) {
+                    Stat stat = fileSystems.getReferenceFileSystem(owner).stat(path.get());
+                    if (isDirectory && stat.fileProperties().isDirectory) {
+                        return path;
+                    } else if (! isDirectory && ! stat.fileProperties().isDirectory) {
+                        return path;
+                    }
+                }
+                keepTrying++;
+            }
+            return Optional.empty();
+        };
+        BiFunction<String, String, Optional<Path>> randomSharedDirectoryPath = (owner, sharee) ->
+             randomSharedPathWithRetries.apply(owner, sharee, true);
+
+        BiFunction<String, String, Optional<Path>> randomSharedFilePath = (owner, sharee) ->
+             randomSharedPathWithRetries.apply(owner, sharee, false);
 
 
         switch (simulation) {
             case READ_OWN_FILE:
                 Path readPath = randomFilePathForUser.get();
                 log(user, simulation, readPath);
-                read(user, readPath);
+                readFile(user, readPath);
                 break;
             case READ_SHARED_FILE:
-                Optional<Path> sharedOpt = randomSharedPath.apply(otherUser.get(), user);
+                Optional<Path> sharedOpt = randomSharedFilePath.apply(user, otherUser.get());
                 if (! sharedOpt.isPresent())
                     return;
                 Path sharedPathToRead = sharedOpt.get();
                 log(user, Simulation.READ_SHARED_FILE, sharedPathToRead);
-                read(user, sharedPathToRead);
+                readFile(user, sharedPathToRead);
+                break;
+            case READ_SHARED_DIRECTORY:
+                Optional<Path> sharedDirOpt = randomSharedDirectoryPath.apply(user, otherUser.get());
+                if (! sharedDirOpt.isPresent())
+                    return;
+                Path sharedDirPathToRead = sharedDirOpt.get();
+                log(user, Simulation.READ_SHARED_DIRECTORY, sharedDirPathToRead);
+                readDirectory(user, sharedDirPathToRead);
                 break;
             case WRITE_OWN_FILE:
                 Path path = getAvailableFilePath(user);
                 log(user, Simulation.WRITE_OWN_FILE, path);
                 write(user, path);
                 break;
             case WRITE_SHARED_FILE:
-                Optional<Path> sharedOpt2 = randomSharedPath.apply(otherUser.get(), user);
+                Optional<Path> sharedOpt2 = randomSharedFilePath.apply(user, otherUser.get());
                 if (! sharedOpt2.isPresent())
                     return;
                 Path  sharedPathToWrite = sharedOpt2.get();
                 log(user, Simulation.WRITE_SHARED_FILE, sharedPathToWrite);
                 write(user, sharedPathToWrite);
                 break;
+            case WRITE_SHARED_DIRECTORY:
```

Comment:
Can we call this `MKDIR_SHARED` instead, please?

Reply:
Write is the terminology unix uses. We are just using mkdir to test it